File: sdk/go/sidecar.go
package gameflow

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"slices"
	"strings"
	"time"
)

// grpcOutOfRange is the gRPC OUT_OF_RANGE code: a list mutation rejected because
// the list is at capacity.
const grpcOutOfRange = 11

// sidecarTransport talks REST to the local GameFlow runtime over plain HTTP.
//
// In v1 the SDK talks to a local sidecar; this is an implementation detail and
// nothing here leaks into the public API, logs or errors.
type sidecarTransport struct {
	client         *http.Client
	watchClient    *http.Client
	baseURL        string
	requestTimeout time.Duration
	logger         Logger
}

func newSidecarTransport(baseURL string, requestTimeout time.Duration, logger Logger) *sidecarTransport {
	return &sidecarTransport{
		// Per-request timeouts are applied via context, so the unary client has
		// no global timeout. The watch client must stream indefinitely.
		client:         &http.Client{},
		watchClient:    &http.Client{},
		baseURL:        baseURL,
		requestTimeout: requestTimeout,
		logger:         logger,
	}
}

// url joins the base URL and a path by plain concatenation. Paths contain custom
// verbs with a literal ':' (lists/players:addValue); concatenation leaves the
// colon untouched.
func (t *sidecarTransport) url(path string) string { return t.baseURL + path }

// do sends a unary request with the per-request timeout. Statuses in `expected`
// are returned to the caller instead of being mapped to an error; any other
// non-2xx becomes a REQUEST_FAILED error.
func (t *sidecarTransport) do(ctx context.Context, method, path string, body any, what string, expected ...int) (*http.Response, error) {
	reqCtx, cancel := context.WithTimeout(ctx, t.requestTimeout)
	// The caller reads the body, so cancel must outlive this function for
	// success/expected responses; it is invoked once the body is closed.
	var reader io.Reader
	if body != nil {
		encoded, err := json.Marshal(body)
		if err != nil {
			cancel()
			return nil, requestFailed(fmt.Sprintf("could not encode %s request", what), 0, err)
		}
		reader = bytes.NewReader(encoded)
	}
	req, err := http.NewRequestWithContext(reqCtx, method, t.url(path), reader)
	if err != nil {
		cancel()
		return nil, requestFailed(fmt.Sprintf("could not build %s request", what), 0, err)
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	res, err := t.client.Do(req)
	if err != nil {
		cancel()
		return nil, t.unreachable(what, err)
	}
	status := res.StatusCode
	if isSuccess(status) || slices.Contains(expected, status) {
		// Hand the response (and a cancel that frees the request context once the
		// body is closed) to the caller.
		res.Body = &cancelOnClose{ReadCloser: res.Body, cancel: cancel}
		return res, nil
	}
	gwErr := readGatewayError(res)
	cancel()
	msg := gwErr.Message
	if msg == "" {
		msg = fmt.Sprintf("%s failed with HTTP %d", what, status)
	}
	return nil, requestFailed(msg, status, nil)
}

func (t *sidecarTransport) unreachable(what string, cause error) *Error {
	return sidecarUnavailable(
		fmt.Sprintf("could not reach the GameFlow runtime at %s (%s)", t.baseURL, what),
		cause,
	)
}

func (t *sidecarTransport) ready(ctx context.Context) error {
	res, err := t.do(ctx, http.MethodPost, "/ready", struct{}{}, "POST /ready")
	if err != nil {
		return err
	}
	drainClose(res)
	return nil
}

func (t *sidecarTransport) health(ctx context.Context) error {
	res, err := t.do(ctx, http.MethodPost, "/health", struct{}{}, "POST /health")
	if err != nil {
		return err
	}
	drainClose(res)
	return nil
}

func (t *sidecarTransport) shutdown(ctx context.Context) error {
	res, err := t.do(ctx, http.MethodPost, "/shutdown", struct{}{}, "POST /shutdown")
	if err != nil {
		return err
	}
	drainClose(res)
	return nil
}

func (t *sidecarTransport) getGameServer(ctx context.Context) (*rawGameServer, error) {
	res, err := t.do(ctx, http.MethodGet, "/gameserver", nil, "GET /gameserver")
	if err != nil {
		return nil, err
	}
	defer drainClose(res)
	var gs rawGameServer
	if err := json.NewDecoder(res.Body).Decode(&gs); err != nil {
		return nil, requestFailed("could not decode the gameserver response", 0, err)
	}
	return &gs, nil
}

func (t *sidecarTransport) getPlayerList(ctx context.Context) (playerListSnapshot, error) {
	res, err := t.do(ctx, http.MethodGet, "/v1beta1/lists/"+playersList, nil, "GET players list", http.StatusNotFound)
	if err != nil {
		return playerListSnapshot{}, err
	}
	defer drainClose(res)
	if res.StatusCode == http.StatusNotFound {
		return disabledSnapshot(), nil
	}
	var list rawList
	if err := json.NewDecoder(res.Body).Decode(&list); err != nil {
		return playerListSnapshot{}, requestFailed("could not decode the players list", 0, err)
	}
	return parseList(&list), nil
}

func (t *sidecarTransport) addPlayer(ctx context.Context, sessionID string) (playerListSnapshot, error) {
	res, err := t.do(
		ctx, http.MethodPost, "/v1beta1/lists/"+playersList+":addValue",
		map[string]string{"value": sessionID}, "addValue",
		http.StatusBadRequest, http.StatusNotFound, http.StatusConflict,
	)
	if err != nil {
		return playerListSnapshot{}, err
	}

	if isSuccess(res.StatusCode) {
		// Some runtime versions answer mutations with an empty/default List
		// instead of the updated one; re-read the list rather than trust it.
		drainClose(res)
		return t.getPlayerList(ctx)
	}

	status := res.StatusCode
	gwErr := readGatewayError(res)
	drainClose(res)
	switch status {
	case http.StatusConflict:
		return playerListSnapshot{}, playerAlreadyConnected(sessionID)
	case http.StatusNotFound:
		return playerListSnapshot{}, playerTrackingDisabled()
	default:
		// 400 OUT_OF_RANGE means the list is at capacity.
		if gwErr.Code == grpcOutOfRange || indicatesCapacity(gwErr.Message) {
			return playerListSnapshot{}, serverFull(0, false)
		}
		msg := gwErr.Message
		if msg == "" {
			msg = fmt.Sprintf("addValue failed with HTTP %d", status)
		}
		return playerListSnapshot{}, requestFailed(msg, status, nil)
	}
}

func (t *sidecarTransport) removePlayer(ctx context.Context, sessionID string) (bool, playerListSnapshot, error) {
	res, err := t.do(
		ctx, http.MethodPost, "/v1beta1/lists/"+playersList+":removeValue",
		map[string]string{"value": sessionID}, "removeValue", http.StatusNotFound,
	)
	if err != nil {
		return false, playerListSnapshot{}, err
	}
	if res.StatusCode == http.StatusNotFound {
		drainClose(res)
		return false, playerListSnapshot{}, nil
	}
	drainClose(res)
	snap, err := t.getPlayerList(ctx)
	if err != nil {
		return false, playerListSnapshot{}, err
	}
	return true, snap, nil
}

func (t *sidecarTransport) watchGameServer(ctx context.Context, sink watchSink) error {
	// Long-lived stream: no per-request timeout, only ctx cancellation.
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, t.url("/watch/gameserver"), nil)
	if err != nil {
		return t.unreachable("GET /watch/gameserver", err)
	}
	res, err := t.watchClient.Do(req)
	if err != nil {
		return t.unreachable("GET /watch/gameserver", err)
	}
	defer drainClose(res)
	if !isSuccess(res.StatusCode) {
		return requestFailed(fmt.Sprintf("watch failed with HTTP %d", res.StatusCode), res.StatusCode, nil)
	}

	reader := bufio.NewReader(res.Body)
	for {
		line, err := reader.ReadBytes('\n')
		if len(line) > 0 {
			t.handleWatchLine(bytes.TrimRight(line, "\r\n"), sink)
		}
		if err != nil {
			if ctx.Err() != nil {
				return nil // cancelled by the caller
			}
			if err == io.EOF {
				return nil // stream ended; caller reconnects
			}
			return t.unreachable("watch stream", err)
		}
	}
}

func (t *sidecarTransport) handleWatchLine(line []byte, sink watchSink) {
	if len(bytes.TrimSpace(line)) == 0 {
		return
	}
	var parsed watchLine
	if err := json.Unmarshal(line, &parsed); err != nil {
		t.logger.Warn("watch: skipping malformed stream line")
		return
	}
	if parsed.Error != nil {
		msg := parsed.Error.Message
		if msg == "" {
			msg = "unknown"
		}
		t.logger.Warn("watch: stream error: " + msg)
		return
	}
	if parsed.Result != nil {
		sink(parsed.Result)
	}
}

func indicatesCapacity(message string) bool {
	lower := strings.ToLower(message)
	return strings.Contains(lower, "capacity") || strings.Contains(lower, "out of range")
}

func readGatewayError(res *http.Response) gatewayError {
	var gwErr gatewayError
	_ = json.NewDecoder(res.Body).Decode(&gwErr)
	return gwErr
}

func isSuccess(status int) bool { return status >= 200 && status < 300 }

// cancelOnClose invokes cancel once the response body is closed, freeing the
// per-request timeout context exactly when the caller is done reading.
type cancelOnClose struct {
	io.ReadCloser
	cancel context.CancelFunc
}

func (c *cancelOnClose) Close() error {
	err := c.ReadCloser.Close()
	c.cancel()
	return err
}

// drainClose drains and closes a response body so the HTTP connection can be
// reused, and releases the request context.
func drainClose(res *http.Response) {
	if res == nil || res.Body == nil {
		return
	}
	_, _ = io.Copy(io.Discard, res.Body)
	_ = res.Body.Close()
}

File: sdk/rust/gameflow/src/transport/sidecar.rs
//! REST transport against the local GameFlow runtime.
//!
//! In v1 the SDK talks to a local sidecar over plain HTTP. This is an
//! implementation detail: nothing here leaks into the public API, logs or
//! errors.

use std::sync::Arc;
use std::time::Duration;

use async_trait::async_trait;
use futures_util::StreamExt;
use reqwest::{Client, RequestBuilder, Response};
use serde_json::json;
use tokio_util::sync::CancellationToken;

use super::{parse_list, Transport, WatchSink, PLAYERS_LIST};
use crate::error::{BoxError, GameFlowError, Result};
use crate::logger::Logger;
use crate::types::{GatewayError, PlayerListSnapshot, RawGameServer, RawList, WatchLine};

/// gRPC `OUT_OF_RANGE`: the list mutation was rejected because it is at capacity.
const GRPC_OUT_OF_RANGE: i64 = 11;

pub(crate) struct SidecarTransport {
    client: Client,
    base_url: String,
    request_timeout: Duration,
    logger: Arc<dyn Logger>,
}

impl SidecarTransport {
    pub(crate) fn new(
        base_url: String,
        request_timeout: Duration,
        logger: Arc<dyn Logger>,
    ) -> Result<Self> {
        // No global timeout: unary calls get a per-request timeout below, while
        // the watch stream must run indefinitely.
        let client = Client::builder().build().map_err(|e| {
            GameFlowError::sidecar_unavailable_with("could not build HTTP client", Some(boxed(e)))
        })?;
        Ok(SidecarTransport {
            client,
            base_url,
            request_timeout,
            logger,
        })
    }

    fn url(&self, path: &str) -> String {
        // Paths contain custom verbs with a literal ':' (lists/players:addValue);
        // the URL is built by plain concatenation and the `url` crate leaves the
        // colon untouched in a path segment.
        format!("{}{}", self.base_url, path)
    }

    /// Sends a unary request with the per-request timeout. Statuses listed in
    /// `expected` are returned to the caller instead of mapping to an error.
    async fn send(&self, req: RequestBuilder, what: &str, expected: &[u16]) -> Result<Response> {
        let res = req
            .timeout(self.request_timeout)
            .send()
            .await
            .map_err(|e| self.unreachable(what, e))?;
        let status = res.status().as_u16();
        if res.status().is_success() || expected.contains(&status) {
            return Ok(res);
        }
        let body = read_gateway_error(res).await;
        Err(GameFlowError::RequestFailed {
            message: body
                .message
                .unwrap_or_else(|| format!("{what} failed with HTTP {status}")),
            status: Some(status),
            source: None,
        })
    }

    fn unreachable(&self, what: &str, cause: reqwest::Error) -> GameFlowError {
        GameFlowError::sidecar_unavailable_with(
            format!(
                "could not reach the GameFlow runtime at {} ({what})",
                self.base_url
            ),
            Some(boxed(cause)),
        )
    }
}

#[async_trait]
impl Transport for SidecarTransport {
    async fn ready(&self) -> Result<()> {
        self.send(
            self.client.post(self.url("/ready")).json(&json!({})),
            "POST /ready",
            &[],
        )
        .await?;
        Ok(())
    }

    async fn health(&self) -> Result<()> {
        self.send(
            self.client.post(self.url("/health")).json(&json!({})),
            "POST /health",
            &[],
        )
        .await?;
        Ok(())
    }

    async fn shutdown(&self) -> Result<()> {
        self.send(
            self.client.post(self.url("/shutdown")).json(&json!({})),
            "POST /shutdown",
            &[],
        )
        .await?;
        Ok(())
    }

    async fn get_game_server(&self) -> Result<RawGameServer> {
        let res = self
            .send(
                self.client.get(self.url("/gameserver")),
                "GET /gameserver",
                &[],
            )
            .await?;
        res.json::<RawGameServer>()
            .await
            .map_err(|e| GameFlowError::RequestFailed {
                message: "could not decode the gameserver response".to_string(),
                status: None,
                source: Some(boxed(e)),
            })
    }

    async fn get_player_list(&self) -> Result<PlayerListSnapshot> {
        let path = format!("/v1beta1/lists/{PLAYERS_LIST}");
        let res = self
            .send(self.client.get(self.url(&path)), "GET players list", &[404])
            .await?;
        if res.status().as_u16() == 404 {
            return Ok(PlayerListSnapshot::disabled());
        }
        let list = res
            .json::<RawList>()
            .await
            .map_err(|e| GameFlowError::RequestFailed {
                message: "could not decode the players list".to_string(),
                status: None,
                source: Some(boxed(e)),
            })?;
        Ok(parse_list(Some(&list)))
    }

    async fn add_player(&self, session_id: &str) -> Result<PlayerListSnapshot> {
        let path = format!("/v1beta1/lists/{PLAYERS_LIST}:addValue");
        let res = self
            .send(
                self.client
                    .post(self.url(&path))
                    .json(&json!({ "value": session_id })),
                "addValue",
                &[400, 404, 409],
            )
            .await?;

        if res.status().is_success() {
            // Some runtime versions answer mutations with an empty/default List
            // instead of the updated one; re-read the list rather than trust it.
            return self.get_player_list().await;
        }

        let status = res.status().as_u16();
        let body = read_gateway_error(res).await;
        match status {
            409 => Err(GameFlowError::PlayerAlreadyConnected {
                session_id: session_id.to_string(),
            }),
            404 => Err(GameFlowError::PlayerTrackingDisabled),
            _ => {
                // 400 OUT_OF_RANGE means the list is at capacity.
                let message = body.message.unwrap_or_default();
                if body.code == Some(GRPC_OUT_OF_RANGE) || indicates_capacity(&message) {
                    Err(GameFlowError::ServerFull { capacity: None })
                } else {
                    Err(GameFlowError::RequestFailed {
                        message: if message.is_empty() {
                            format!("addValue failed with HTTP {status}")
                        } else {
                            message
                        },
                        status: Some(status),
                        source: None,
                    })
                }
            }
        }
    }

    async fn remove_player(&self, session_id: &str) -> Result<Option<PlayerListSnapshot>> {
        let path = format!("/v1beta1/lists/{PLAYERS_LIST}:removeValue");
        let res = self
            .send(
                self.client
                    .post(self.url(&path))
                    .json(&json!({ "value": session_id })),
                "removeValue",
                &[404],
            )
            .await?;
        if res.status().as_u16() == 404 {
            return Ok(None);
        }
        Ok(Some(self.get_player_list().await?))
    }

    async fn watch_game_server(&self, sink: WatchSink, cancel: CancellationToken) -> Result<()> {
        // Long-lived stream: no per-request timeout, only cancellation.
        let res = self
            .client
            .get(self.url("/watch/gameserver"))
            .send()
            .await
            .map_err(|e| self.unreachable("GET /watch/gameserver", e))?;
        if !res.status().is_success() {
            return Err(GameFlowError::RequestFailed {
                message: format!("watch failed with HTTP {}", res.status().as_u16()),
                status: Some(res.status().as_u16()),
                source: None,
            });
        }

        let mut stream = res.bytes_stream();
        let mut buffer: Vec<u8> = Vec::new();
        loop {
            tokio::select! {
                _ = cancel.cancelled() => return Ok(()),
                chunk = stream.next() => {
                    match chunk {
                        None => return Ok(()), // stream ended; caller reconnects
                        Some(Err(e)) => return Err(self.unreachable("watch stream", e)),
                        Some(Ok(bytes)) => {
                            buffer.extend_from_slice(&bytes);
                            // Lines may be split across chunks; frame by newline.
                            while let Some(pos) = buffer.iter().position(|&b| b == b'\n') {
                                let line: Vec<u8> = buffer.drain(..=pos).collect();
                                let mut line = &line[..line.len() - 1];
                                if line.last() == Some(&b'\r') {
                                    line = &line[..line.len() - 1];
                                }
                                if line.is_empty() {
                                    continue;
                                }
                                self.handle_watch_line(line, &sink);
                            }
                        }
                    }
                }
            }
        }
    }
}

impl SidecarTransport {
    fn handle_watch_line(&self, line: &[u8], sink: &WatchSink) {
        let parsed: WatchLine = match serde_json::from_slice(line) {
            Ok(parsed) => parsed,
            Err(_) => {
                self.logger.warn("watch: skipping malformed stream line");
                return;
            }
        };
        if let Some(error) = parsed.error {
            self.logger.warn(&format!(
                "watch: stream error: {}",
                error.message.unwrap_or_else(|| "unknown".to_string())
            ));
            return;
        }
        if let Some(gs) = parsed.result {
            sink(gs);
        }
    }
}

fn indicates_capacity(message: &str) -> bool {
    let lower = message.to_ascii_lowercase();
    lower.contains("capacity") || lower.contains("out of range")
}

async fn read_gateway_error(res: Response) -> GatewayError {
    res.json::<GatewayError>().await.unwrap_or_default()
}

fn boxed<E: std::error::Error + Send + Sync + 'static>(error: E) -> BoxError {
    Box::new(error)
}
